File: frontend/src/types/mission.ts
export type C2ConnectionMode = 'SETUP_MODE' | 'OPS_MONITOR_MODE' | 'PRAMAAN_CONTROL_MODE' | 'FUTURE_SECURE_CONTROL_MODE';
export type MAVLinkBridgeStatus = 'NOT_CONFIGURED' | 'SIMULATION_ONLY' | 'READ_ONLY_READY' | 'CONNECTED_READ_ONLY' | 'ERROR';
export type MissionDraftStatus = 'DRAFT' | 'VALIDATED' | 'INVALID' | 'LOCKED';
export type WaypointAction = 'NAVIGATE' | 'LOITER' | 'CAPTURE_IMAGE' | 'START_RECORDING' | 'STOP_RECORDING' | 'RETURN_POINT';

export interface BridgeStatus { status: MAVLinkBridgeStatus; read_only: boolean; endpoint_count: number; message: string; hardware_commands_enabled: boolean; }
export interface MAVLinkEndpoint { id: number; endpoint_id: string; name: string; host: string; port: number; protocol: string; status: MAVLinkBridgeStatus; read_only: boolean; created_at: string; updated_at: string; }
export interface C2ConnectionConfig { id?: number; mode: C2ConnectionMode; description: string; mission_planner_allowed: boolean; pramaan_commands_allowed: boolean; hardware_commands_enabled: boolean; puf_required: boolean; created_at?: string; updated_at?: string; }

export interface MissionDraft { id: number; mission_id: string; name: string; drone_id: string; vehicle_type: 'QUADCOPTER' | 'FIXED_WING'; status: MissionDraftStatus; default_altitude_m: number; default_speed_mps: number; lost_link_action: string; created_at: string; updated_at: string; }
export interface MissionDraftCreate { name: string; drone_id: string; vehicle_type: 'QUADCOPTER' | 'FIXED_WING'; default_altitude_m: number; default_speed_mps: number; lost_link_action: string; }
export interface MapWaypoint { id: number; mission_id: string; sequence: number; lat: number; lon: number; altitude_m: number; speed_mps: number; action: WaypointAction; loiter_seconds?: number | null; notes?: string | null; }
export interface MapWaypointCreate { lat: number; lon: number; altitude_m?: number; speed_mps?: number; action: WaypointAction; loiter_seconds?: number | null; notes?: string | null; }
export interface MissionValidation { mission_id: string; status: MissionDraftStatus; valid: boolean; errors: string[]; warnings: string[]; }
export interface GeofenceDraft { id: number; geofence_id: string; name: string; drone_id: string; enabled: boolean; polygon_json: string; max_altitude_m: number; min_altitude_m: number; created_at: string; updated_at: string; }

File: backend/app/models/mission.py
from datetime import datetime, timezone

from sqlalchemy import Boolean, DateTime, Float, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.core.enums import MissionDraftStatus, WaypointAction
from app.db.session import Base


class MissionDraft(Base):
    __tablename__ = "mission_drafts"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    mission_id: Mapped[str] = mapped_column(String(64), unique=True, index=True)
    name: Mapped[str] = mapped_column(String(128))
    drone_id: Mapped[str] = mapped_column(String(64), index=True)
    vehicle_type: Mapped[str] = mapped_column(String(32))
    status: Mapped[str] = mapped_column(String(32), default=MissionDraftStatus.DRAFT.value)
    default_altitude_m: Mapped[float] = mapped_column(Float, default=50.0)
    default_speed_mps: Mapped[float] = mapped_column(Float, default=8.0)
    lost_link_action: Mapped[str] = mapped_column(String(64), default="HOLD_THEN_RTL")
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))


class MapWaypoint(Base):
    __tablename__ = "map_waypoints"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    mission_id: Mapped[str] = mapped_column(String(64), index=True)
    sequence: Mapped[int] = mapped_column(Integer)
    lat: Mapped[float] = mapped_column(Float)
    lon: Mapped[float] = mapped_column(Float)
    altitude_m: Mapped[float] = mapped_column(Float)
    speed_mps: Mapped[float] = mapped_column(Float)
    action: Mapped[str] = mapped_column(String(32), default=WaypointAction.NAVIGATE.value)
    loiter_seconds: Mapped[int | None] = mapped_column(Integer, nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)


class GeofenceDraft(Base):
    __tablename__ = "geofence_drafts"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    geofence_id: Mapped[str] = mapped_column(String(64), unique=True, index=True)
    name: Mapped[str] = mapped_column(String(128))
    drone_id: Mapped[str] = mapped_column(String(64), index=True)
    enabled: Mapped[bool] = mapped_column(Boolean, default=True)
    polygon_json: Mapped[str] = mapped_column(Text)
    max_altitude_m: Mapped[float] = mapped_column(Float)
    min_altitude_m: Mapped[float] = mapped_column(Float, default=0.0)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))


class MissionEvent(Base):
    __tablename__ = "mission_events"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    mission_id: Mapped[str] = mapped_column(String(64), index=True)
    drone_id: Mapped[str] = mapped_column(String(64), index=True)
    event_type: Mapped[str] = mapped_column(String(64), index=True)
    severity: Mapped[str] = mapped_column(String(16), default="INFO")
    message: Mapped[str] = mapped_column(String(512))
    timestamp: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), index=True)
    details: Mapped[str | None] = mapped_column(Text, nullable=True)
